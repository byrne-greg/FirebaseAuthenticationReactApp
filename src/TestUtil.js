import React from "react";
import { MemoryRouter } from "react-router-dom";

const withTestRouter = Component => <MemoryRouter>{Component}</MemoryRouter>;

export { withTestRouter };
