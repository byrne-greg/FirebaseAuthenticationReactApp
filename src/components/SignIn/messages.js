export default {
  signIn: "Sign In",
  emailAddressLabel: "Email Address",
  email: "email",
  passwordLabel: "Password",
  password: "password"
};
