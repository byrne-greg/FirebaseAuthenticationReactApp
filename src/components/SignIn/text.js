export default {
  signIn: "Sign In",
  emailAddress: "Email Address",
  password: "Password"
};
