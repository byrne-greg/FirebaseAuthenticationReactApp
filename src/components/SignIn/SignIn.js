import React, { useState, useContext } from "react";
import { withRouter } from "react-router-dom";

import { SignUpLink } from "../SignUp";
import { PasswordForgetLink } from "../PasswordForget";
import { FirebaseContext } from "../Firebase";
import * as ROUTES from "../../constants/routes";
import text from "./text";

const SignIn = () => (
  <div>
    <h1>{text.signIn}</h1>
    <SignInForm />
    <PasswordForgetLink />
    <SignUpLink />
  </div>
);

const ErrorAlert = ({ errorText }) => (
  <div id="submit-error">
    <p>{errorText}</p>
  </div>
);

const SubmitButton = ({ isInvalid }) => (
  <button id="submit-button" disabled={isInvalid} type="submit">
    {text.signIn}
  </button>
);

const EmailInput = ({ currentValue, onChange }) => (
  <>
    <label id="email-label">{text.emailAddress}</label>
    <input
      aria-labelledby="email-label"
      id="email-input"
      name="email"
      value={currentValue}
      onChange={onChange}
      type="email"
      placeholder={text.emailAddress}
    />
  </>
);

const PasswordInput = ({ currentValue, onChange }) => (
  <>
    <label id="password-label">{text.password}</label>
    <input
      aria-labelledby="password-label"
      id="password-input"
      name="password"
      value={currentValue}
      onChange={onChange}
      type="password"
      placeholder={text.password}
    />
  </>
);

const SignInFormComponent = ({ history }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState(null);
  const firebase = useContext(FirebaseContext);

  const isInvalid = !password || !email;

  const onSubmit = async event => {
    event.preventDefault();
    try {
      await firebase.login(email, password);
      setEmail("");
      setPassword("");
      history.push(ROUTES.HOME);
    } catch (error) {
      setError({ ...error });
    }
  };

  const handleChange = ({ event, setState }) => {
    setState(event.target.value);
    setError(null);
  };
  const onEmailChange = event => {
    handleChange({ event, setState: setEmail });
  };
  const onPasswordChange = event => {
    handleChange({ event, setState: setPassword });
  };

  return (
    <form onSubmit={onSubmit}>
      <EmailInput value={email} onChange={onEmailChange} />
      <PasswordInput value={password} onChange={onPasswordChange} />
      <SubmitButton isInvalid={isInvalid} />
      {error && <ErrorAlert errorText={error.message} />}
    </form>
  );
};
const SignInForm = withRouter(SignInFormComponent);

export default SignIn;
export { SignInForm };
