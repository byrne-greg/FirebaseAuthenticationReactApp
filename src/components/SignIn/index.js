import SignIn, { SignInForm } from "./SignIn";

export default SignIn;
export { SignInForm };
