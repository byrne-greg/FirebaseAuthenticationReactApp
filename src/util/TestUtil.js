import React from "react";
import { MemoryRouter } from "react-router-dom";
import { FirebaseContext } from "../components/Firebase";
import { AuthUserContext } from "../components/Session";

const withMockFirebase = (mockFirebase, Component) => (
  <FirebaseContext.Provider value={mockFirebase}>
    {Component}
  </FirebaseContext.Provider>
);

const withMockAuthUserContext = (value, Component) => (
  <AuthUserContext.Provider value={value}>{Component}</AuthUserContext.Provider>
);

const withTestRouter = Component => <MemoryRouter>{Component}</MemoryRouter>;

export { withTestRouter, withMockFirebase, withMockAuthUserContext };
