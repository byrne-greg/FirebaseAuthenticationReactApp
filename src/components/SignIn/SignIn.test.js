import React from "react";
import { shallow, mount } from "enzyme";
import { withTestRouter } from "../../TestUtil";
import SignIn, { SignInForm } from ".";

const setState = jest.fn();
const useStateSpy = jest.spyOn(React, "useState");
useStateSpy.mockImplementation(init => [init, setState]);

describe("SignIn render tests", () => {
  it("page renders as expected", () => {
    // GIVEN

    // WHEN
    const page = mount(withTestRouter(<SignIn />));

    // THEN
    expect(page).toMatchSnapshot();
  });

  it("form submit button is disabled when input fields have no text", () => {
    // GIVEN

    // WHEN
    const form = mount(withTestRouter(<SignInForm.WrappedComponent />));
    const submitButton = form.find("#submit-button").props();

    // THEN
    expect(submitButton.disabled).toStrictEqual(true);
  });

  xit("form submit button is enabled when input fields have text", () => {
    // // GIVEN
    // // WHEN
    // let form;
    // act(() => {
    //   form = mount(withTestRouter(<SignInForm.WrappedComponent />));
    // });
    // const emailField = form.find("#email-input-field");
    // emailField.props().value = "email";
    // act(() => {
    //   emailField.props().onChange({ target: { value: "new email" } });
    // });
    // const passwordField = form.find("#password-input-field");
    // passwordField.props().value = "password";
    // act(() => {
    //   passwordField.props().onChange({ target: { value: "new password" } });
    // });
    // const submitButton = form.find("#submit-button").props();
    // // THEN
    // expect(submitButton.disabled).toStrictEqual(false);
    // // expect(form).toMatchSnapshot();
  });

  xit("form renders with error", () => {
    // GIVEN

    // WHEN
    const form = mount(withTestRouter(<SignInForm.WrappedComponent />));

    // THEN
    expect(form).toMatchSnapshot();
  });
});

describe("SignIn unit tests", () => {
  xit("email state changes when email field has text", () => {
    // // GIVEN
    // const signInForm = shallow(<SignInForm />);
    // // WHEN
    // const email = signInForm.find("input").props();
    // // THEN
    // expect(email).toStrictEqual("a");
  });
  xit("password state changes when password field has text", () => {
    // // GIVEN
    // // WHEN
    // // THEN
  });
  xit("login is attempted when submit button is pressed", () => {
    // // GIVEN
    // // WHEN
    // // THEN
  });
});
