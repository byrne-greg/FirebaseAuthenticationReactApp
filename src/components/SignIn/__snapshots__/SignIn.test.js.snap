// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn render tests form renders with error 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    id="email-input-field"
    name="email"
    onChange={[Function]}
    placeholder="Email Address"
    type="text"
  />
  <input
    id="password-input-field"
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="text"
  />
  <button
    disabled={true}
    id="submit-button"
    type="submit"
  >
    Sign In
  </button>
</form>
`;

exports[`SignIn render tests form submit button is enabled when input fields have text 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    id="email-input-field"
    name="email"
    onChange={[Function]}
    placeholder="Email Address"
    type="text"
  />
  <input
    id="password-input-field"
    name="password"
    onChange={[Function]}
    placeholder="Password"
    type="text"
  />
  <button
    disabled={true}
    id="submit-button"
    type="submit"
  >
    Sign In
  </button>
</form>
`;

exports[`SignIn render tests page renders as expected 1`] = `
<div>
  <h1>
    SignIn
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      id="email-input-field"
      name="email"
      onChange={[Function]}
      placeholder="Email Address"
      type="text"
    />
    <input
      id="password-input-field"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="text"
    />
    <button
      disabled={true}
      id="submit-button"
      type="submit"
    >
      Sign In
    </button>
  </form>
  <p>
    <a
      href="/pw-forget"
      onClick={[Function]}
    >
      Forgot Password?
    </a>
  </p>
  <p>
    Don't have an account? 
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign Up
    </a>
  </p>
</div>
`;
